doc/source/user/01_configurations.rst
.. _configuration-examples:

Configuration examples
======================

Configuration location is ``${XDG_CONFIG_HOME}/mpd_sima/mpd_sima.cfg`` usually ``${XDG_CONFIG_HOME}``  is in ``${HOME}/.config/``.


Default configuration
---------------------

MPD_sima runs without explicit configuration as long as your MPD server is easy
to discover (``localhost:6600`` or exposed via environment variables ``MPD_HOST``/``MPD_PORT``).

The default configuration is to run in *track mode* with suggestions from
last.fm. In addition to `Lastfm` plugin set in **[sima]** section (within *internal*
option) there is a `Crop` plugin (keeps 10 tracks behind the currently playing
track) and a `Random` plugin (acting as a fallback if `Lastfm` did not find any
track to queue).

.. code:: ini

    [MPD]
    # Change MPD server here
    #host = mpdserver.example.org
    #port = 6600

    [sima]
    internal = Crop, Lastfm, Random
    history_duration = 8
    queue_length = 2

    [crop]
    consume = 10

    [lastfm]
    queue_mode = track
    single_album = False
    track_to_add = 1

Album mode
^^^^^^^^^^

One of the first request added to MPD_sima was album mode. It allows to queue whole album instead of single tracks.

Here the configuration keeps the queue plugin *Lastfm* but configures it to queue albums (``queue_mode = album``) and ask for 2 albums to be add.

The configuration of MPD_sima in ``sima`` section is also modified for ``queue_length``. The value of 10 is to trigger album queueing when there are 10 tracks or less in the queue.

.. code:: ini

    [sima]
    internal = Crop, Lastfm, Random
    history_duration = 24
    queue_length = 10

    [crop]
    consume = 10

    [lastfm]
    queue_mode = album
    album_to_add = 2


Offline auto-queuing
--------------------

In addition to LastFm there are other *plugins* you can mix together or use alone to have a specific queue mode.

Genre auto-queuing
^^^^^^^^^^^^^^^^^^
With this mode MPD_sima is trying to queue tracks with genre tag similar to previously played tracks.

.. code:: ini

    [sima]
    internal = Crop, Genre, Random
    history_duration = 8

    [genre]
    queue_mode = track
    track_to_add = 1


Random queuing
^^^^^^^^^^^^^^
This mode allows random queuing with different flavour **pure** (total
randomness) or **sensible** (use play history to filter chosen tracks).
The **sensible** flavour will then tend to play track not in recently played.

The history duration is important for this plugin when running in **sensible** flavour since you might exhaust possible tracks to queue if the **history_duration** is larger than the total play time you have in your music library.

.. code:: ini

    [sima]
    internal = Crop, Random
    history_duration = 168

    [random]
    flavour = sensible

Tag queuing
^^^^^^^^^^^

This is the most complex and versatile offline mode. "Tags" plugin allows to queue track based on actual tags value.

Here is an example to have MPD_sima to queue only electronic music tagged with genres **electronica** or **IDM** or **glitch**:

.. code:: ini

    [sima]
    internal = Crop, Tags

    [tags]
    # Look for files with tagged with genre "electronica" OR "IDM" OR "glitch"
    genre = electronica, IDM, glitch

There are other supported tags, mainly **date**, **originaldate** or
**comment** (cf manual for the exact list). You can use more than one, entries
in tags sections are ANDed within a single MPD filter to look for titles.

For instance setting "genre=rock" and
"date=1982,1983,1984,1985,1986,1987,1988,1989" will end up looking for track
tagged with genre rock and date within 1982 through 1989:

.. code:: ini

    [sima]
    internal = Crop, Tags

    [tags]
    genre = rock
    date = 1982,1983,1984,1985,1986,1987,1988,1989

In case you want to make complex search in MPD library you can provide an `MPD
filter`_. This is a powerful feature, it comes at the cost of a different syntax, some might find it more readable some wont :D

For instance in the previous example you can simply replace **date** with the
current filter "`(date =~ '198[2-9]+')`":

.. code:: ini

    [sima]
    internal = Crop, Tags

    [tags]
    genre = rock
    filter = (date =~ '198[2-9]+')

And even go further and merge genre in the filter using "`(genre == 'rock') AND
(date =~ '198[2-9]+')`".

.. code:: ini

    [sima]
    internal = Crop, Tags

    [tags]
    filter = (genre == 'rock' ) AND (date =~ '198[2-9]+')

Since the setup for the filter can be tricky and it can be useful to validate
the syntax and have a look at what kind of artists the filter would return.

You can call MPD_sima to controls the configuration file. For instance with a config in `sima.cfg` run:

.. code:: bash

    ./mpd-sima --log-level info --config sima.cfg config-test


.. include:: ../links.rst

.. vim: spell spelllang=en


doc/Changelog
MPD_sima v0.18.2

  * Add positional argument:
  "random", a short cut to start in random mode

 -- kaliko <kaliko@azylum.org>  UNRELEASED

MPD_sima v0.18.1

 * Remove bad heuristic to infer artist aliases
 * Set max_art default value to 20
 * Add sphinx documentation
 * Merge unix manuals in sphinx reStructuredText
   Call "setup.py build_sphinx" to build documentation (troff and html).
   Build only man pages:
   sphinx-build -d ./build/sphinx/doctrees doc/source -b man ./build/sphinx/man

 -- kaliko <kaliko@azylum.org>  Mon, 07 Feb 2022 18:11:48 +0100

MPD_sima v0.18.0

 * New database, no migration done from old database !
   A backup is done in "${XDG_DATA_HOME}/mpd_sima/"
   Play history and blocklist are forgotten then.
 * Blocklist is now handled with mpd-sima command arguments,
   simadb_cli executable is removed.
 * Removed obsolete --create-db and --generate-config options
 * Fixed crash when setting XDG_CONFIG_HOME (closes #50)
 * Add abstract unix socket support for MPD connection
 * Better MPD connection management.
 * Need at least python-musicpd 0.7.0.
   Abstract socket in v0.6.0 and socket timeout in v0.7.0

 -- kaliko <kaliko@azylum.org>  Tue, 16 Nov 2021 16:02:37 +0100

MPD_sima v0.17.0

 * New offline autoqueue plugin based on file genre tags only
 * Add randomness in Tags plugin artist selection
 * Fixed issue with un-tagged titles (closes #40)
 * Fixed "artist with quotes not found" (closes #41)
 * Fixed bad pid in Uniq core plugin (closes #32)
 * Move "test-tags-config" as positional argument "config-test"
 * Move "--create-db" as positional argument "create-db"
 * Add mpd-sima commands as positional arguments:
 config-test, create-db, generate-config, purge-history
 * Use duration instead of time in Track object (closes #37)
 * Use musicbrainz_albumid|trackid tags in Tracks object

 -- kaliko <kaliko@azylum.org>  Wed, 28 Apr 2021 17:21:39 +0200

MPD_sima v0.16.1

 * tags plugin: Add album queue mode
 * tags plugin: Ensure metadata used in filter are enabled
 so that MPD exposes them (closes #38)

 -- kaliko <kaliko@azylum.org>  Sat, 09 Jan 2021 12:28:17 +0100

MPD_sima v0.16.0

 * New offline autoqueue plugin based on file tags only
 * Expose a new executable "test-tags-config"
 * Update vinstall.py (closes #34)
 * Fixed bug when http client fetches a corrupted cached entry (closes #35)
 * Need at least python 3.6

 -- kaliko <kaliko@azylum.org>  Tue, 15 Dec 2020 16:17:05 +0100

MPD_sima v0.16.0.dev0

 * Major MPD client refactoring
 * Refactored random plugin
 * Fixed bug in MPD client reconnection

 -- kaliko <kaliko@azylum.org>  Sat, 16 May 2020 14:49:04 +0200

MPD_sima v0.15.3

 * Improve HTTP client (use persistent connection)
 * Fixed a serious bug in PlayerClient().search_albums()
   "search_albums" method was ignoring some albums when Artist object had
   aliases.

-- kaliko jack <kaliko@azylum.org> Sun, 03 May 2020 16:30:36 +0200

MPD_sima v0.15.2

 * Fixed sqlite sqlite3.OperationalError VACUUM Error
 cf. https://bugs.python.org/issue28518
 * Add option to queue a chosen number of tracks from an album (closes #28)
 * Add option to shuffle tracks in album mode (Thanks Sacha)

-- kaliko jack <kaliko@azylum.org> Thu, 14 Nov 2019 17:41:22 +0100

MPD_sima v0.15.1

 * Fixed mopidy compatibility implementation

-- kaliko jack <kaliko@azylum.org> Tue, 19 Mar 2019 15:14:48 +0100

MPD_sima v0.15.0

 * Remove EchoNest support (Closes #10)
 * Enhanced queuing behavior in random mode (Closes #16)
 * Add option to prevent single & repeat mode to disable queuing (Closes #19)
 * Honor MPC password/host format on command line option 'host'
 * Fixed fuzzy search for short artist names
 * Client side hardcoded one month caching (ugly workaround for #7)
 * Simplified album search
 * Track search aggregates artist name and MusicBrainzID searches
 * Add an undocumented mopidy compatibility configuration option
   Use the boolean option "mopidy_compat" in the sima section
   cf. https://gitlab.com/kaliko/sima/issues/21#

-- kaliko jack <kaliko@azylum.org> Tue, 19 Mar 2019 13:58:49 +0100

MPD_sima v0.14.4

 * Fixed blacklisting for Random plugin (Closes #14)
 * Fixed sensible flavour of random plugin, thanks dnmx (Closes #15)

-- kaliko jack <kaliko@azylum.org> Thu, 24 Aug 2017 11:46:14 +0200


MPD_sima v0.14.3

 * This release in only intended to get rid of 0.14.2 source tarball.
   Source tar ball for v0.14.2 was badly generated. Release 0.14.2 and 0.14.3
   are actually the same source code, only generated tar were affected.
   Sorry for the inconvenience.

-- kaliko jack <kaliko@azylum.org> Sun, 12 Feb 2017 15:49:03 +0100


MPD_sima v0.14.2

 * Fixed MPD client cache (filters empty strings)
 * Fixed Track object to work with MPD >= 0.20 (Closes #12)
 * Fixed python virtualenv installer

-- kaliko jack <kaliko@azylum.org> Sat, 11 Feb 2017 13:04:27 +0100


MPD_sima v0.14.1

 * Prevent adding empty track to history (Closes #5)
 * Fixed false warning about wrong MBID format (Closes #6)

-- kaliko jack <kaliko@azylum.org> Sat, 02 Jan 2016 16:03:39 +0100


MPD_sima v0.14.0

 * Add option to generate config on stdout
 * Add priority feature for plugins
 * More robust MPD client
 * Fixed top track mode
 * Fixed default conf file name (back to mpd_sima.cfg)
 * Fixed http cache for lastFM
 * Cleanup code (removed users similarities)
 * Make "var-dir" option default name instead of "var_dir"
 * Honor HTTP_PROXY env. var. (need requests >2.4.0)

-- kaliko jack <kaliko@azylum.org> Sat, 17 Oct 2015 10:48:41 +0200


MPD_sima v0.13.0

 * Use of MusicBrainzIdentifier by default
   Please consider tagging your library with these metadata
 * New virtualenv installer in pure python

-- kaliko jack <kaliko@azylum.org>  Wed, 28 Jan 2015 12:20:49 +0100


MPD_sima v0.12.3

 * Use albumartist to fetch similar artists
 * Fixed some bugs managing tags

-- kaliko jack <kaliko@azylum.org>  Sat, 06 Dec 2014 16:42:15 +0100


MPD_sima v0.12.2

 * Add some randomness to track selection

sima/utils/config.py
# -*- coding: utf-8 -*-
# Copyright (c) 2009-2015, 2019-2021 kaliko <kaliko@azylum.org>
# Copyright (c) 2019 sacha <sachahony@gmail.com>
#
#  This file is part of sima
#
#  sima is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  sima is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with sima.  If not, see <http://www.gnu.org/licenses/>.
#

"""
Deal with configuration and data files.
Parse configuration file and set defaults for missing options.
"""

# IMPORTS
import configparser
import logging
import sys

from configparser import Error
from os import (access, makedirs, getenv, stat, chmod, W_OK)
from os.path import (join, isdir, isfile, dirname, exists)
from stat import (S_IMODE, ST_MODE, S_IRWXO, S_IRWXG)

from . import utils

# DEFAULTS
DIRNAME = 'mpd_sima'
CONF_FILE = 'mpd_sima.cfg'

DEFAULT_CONF = {
        'MPD': {
            'host': "localhost",
            #'password': "",
            'port': 6600,
            },
        'sima': {
            'internal': "Crop, Lastfm, Random",
            'contrib': "",
            'user_db': False,
            'history_duration': 8,
            'queue_length': 2,
            'var_dir': 'empty',
            'musicbrainzid': True,
            'repeat_disable_queue': True,
            'single_disable_queue': True,
            'mopidy_compat': False,
            },
        'daemon': {
            'daemon': False,
            'pidfile': "",
            },
        'log': {
            'verbosity': "info",
            'logfile': "",
            },
        'crop': {
            'consume': 10,
            'priority': 0,
            },
        'lastfm': {
            'queue_mode': "track",  # TODO control values
            'max_art': 20,
            'single_album': False,
            'track_to_add': 1,
            'album_to_add': 1,
            'shuffle_album': False,
            'track_to_add_from_album': 0,  # <=0 means keep all
            'depth': 1,
            'cache': True,
            'priority': 100,
            },
        'random': {
            'flavour': "sensible",  # in pure, sensible
            'track_to_add': 1,
            'priority': 50,
            },
        'tags': {
            'comment': "",
            'date': "",
            'genre': "",
            'label': "",
            'originaldate': "",
            'filter': "",
            'queue_mode': "track",
            'single_album': False,
            'track_to_add': 1,
            'album_to_add': 1,
            'priority': 80,
            },
        'genre': {
            'queue_mode': "track",
            'single_album': False,
            'track_to_add': 1,
            'album_to_add': 1,
            'priority': 80,
            },
        }
#


class ConfMan:  # CONFIG MANAGER CLASS
    """
    Configuration manager.
    Default configuration is stored in DEFAULT_CONF dictionnary.
    First init_config() run to get config from file.
    Then control_conf() is run and retrieve configuration from defaults if not
    set in conf files.
    These settings are then updated with command line options with
    supersedes_config_with_cmd_line_options().

    Order of priority for the origin of an option is then (lowest to highest):
        * DEFAULT_CONF
        * Env. Var for MPD host, port and password
        * configuration file (overrides previous)
        * command line options (overrides previous)
    """

    def __init__(self, options=None):
        self.log = logging.getLogger('sima')
        # options settings priority:
        # defauts < env. var. < conf. file < command line
        self.conf_file = options.get('conf_file')
        self.config = configparser.ConfigParser(inline_comment_prefixes='#')
        self.config.read_dict(DEFAULT_CONF)
        # update DEFAULT_CONF with env. var.
        self.use_envar()
        self.startopt = options

        # INIT CALLS
        self.init_config()
        self.supersedes_config_with_cmd_line_options()
        # set dbfile
        self.config['sima']['db_file'] = join(self.config['sima']['var_dir'], 'sima.db')
        # Controls files access
        self.control_facc()

        # Create directories
        data_dir = self.config['sima']['var_dir']
        if not isdir(data_dir):
            self.log.trace('Creating "%s"', data_dir)
            makedirs(data_dir)
            chmod(data_dir, 0o700)

    def control_facc(self):
        """Controls file access.
        This is relevant only for file provided through the configuration file
        since files provided on the command line are already checked with
        argparse. Also add config['sima']['db_file'] contructed here in init
        """
        ok = True
        for op, ftochk in [('logfile', self.config.get('log', 'logfile')),
                           ('pidfile', self.config.get('daemon', 'pidfile')),
                           ('db file', self.config.get('sima', 'db_file'))]:
            if not ftochk:
                continue
            if isdir(ftochk):
                self.log.critical('Need a file not a directory: "%s"', ftochk)
                ok = False
            if not exists(ftochk):
                # Is parent directory writable then
                filedir = dirname(ftochk)
                if not access(filedir, W_OK):
                    self.log.critical('no write access to "%s" (%s)', filedir, op)
                    ok = False
            else:
                if not access(ftochk, W_OK):
                    self.log.critical('no write access to "%s" (%s)', ftochk, op)
                    ok = False
        if not ok:
            sys.exit(2)

    def control_mod(self):
        """
        Controls conf file permissions.
        """
        mode = S_IMODE(stat(self.conf_file)[ST_MODE])
        self.log.debug('file permission is: %o', mode)
        if mode & S_IRWXO or mode & S_IRWXG:
            self.log.warning('File is readable by "other" and/or' +
                             ' "group" (actual permission %o octal).', mode)
            self.log.warning('Consider setting permissions to 600 octal.')

    def supersedes_config_with_cmd_line_options(self):
        """Updates defaults settings with command line options"""
        for sec in self.config.sections():
            for opt in self.config.options(sec):
                if opt in list(self.startopt.keys()):
                    self.config.set(sec, opt, str(self.startopt.get(opt)))
        # honor MPD_HOST format as in mpc(1)  for command line option --host
        if self.startopt.get('host'):
            if '@' in self.startopt.get('host'):
                host, passwd = utils.parse_mpd_host(self.startopt.get('host'))
                if passwd:
                    self.config.set('MPD', 'password', passwd)
                if host:
                    self.config.set('MPD', 'host', host)

    def use_envar(self):
        """Use MPD en.var. to set defaults"""
        mpd_host, mpd_port, passwd = utils.get_mpd_environ()
        if mpd_host:
            self.log.info('Env. variable MPD_HOST set to "%s"', mpd_host)
            self.config['MPD'].update(host=mpd_host)
        if passwd:
            self.log.info('Env. variable MPD_HOST contains password.')
            self.config['MPD'].update(password=passwd)
        if mpd_port:
            self.log.info('Env. variable MPD_PORT set to "%s".', mpd_port)
            self.config['MPD'].update(port=mpd_port)

    def init_config(self):
        """
        Use XDG directory standard if exists
        else use "HOME/(.config|.local/share)/sima/"
        http://standards.freedesktop.org/basedir-spec/basedir-spec-0.6.html
        """

        homedir = getenv('HOME')

        if getenv('XDG_DATA_HOME'):
            data_dir = join(getenv('XDG_DATA_HOME'), DIRNAME)
        elif homedir and isdir(homedir) and homedir not in ['/']:
            data_dir = join(homedir, '.local', 'share', DIRNAME)
        else:
            self.log.critical('Can\'t find a suitable location for data folder (XDG_DATA_HOME)')
            self.log.critical('Please use "--var-dir" to set a proper location')
            sys.exit(1)

        if self.startopt.get('conf_file'):
            # No need to handle conf file location
            pass
        elif getenv('XDG_CONFIG_HOME'):
            conf_dir = join(getenv('XDG_CONFIG_HOME'), DIRNAME)
            self.conf_file = join(conf_dir, CONF_FILE)
        elif homedir and isdir(homedir) and homedir not in ['/']:
            conf_dir = join(homedir, '.config', DIRNAME)
            self.conf_file = join(conf_dir, CONF_FILE)
        else:
            self.log.critical('Can\'t find a suitable location for config folder (XDG_CONFIG_HOME)')
            self.log.critical('Please use "--config" to locate the conf file')
            sys.exit(1)

        # Sima sqlite DB
        self.config['sima']['var_dir'] = join(data_dir)

        # If no conf file present, uses defaults
        if not isfile(self.conf_file):
            return

        self.log.info('Loading configuration from:  %s', self.conf_file)
        self.control_mod()

        try:
            self.config.read(self.conf_file)
        except Error as err:
            self.log.error(err)
            sys.exit(1)

# VIM MODLINE
# vim: ai ts=4 sw=4 sts=4 expandtab
